File: katas/kata-1/fsharp/TimeToActParser.Tests/ComplexScenariosTests.fs
module TimeToActParser.Tests.ComplexScenariosTests

open Xunit
open System.Text.Json
open System.Text.Json.Nodes
open TimeToActParser.DocumentParser
open TimeToActParser.JsonSerializer
open TimeToActParser.Tests.TestDataLoader

type ComplexScenariosTests() =
    static member ComplexScenariosTestData() =
        let testCases = getTestCases "complex-scenarios.json"
        testCases |> Array.map (fun tc -> [| tc |])

    [<Theory>]
    [<MemberData(nameof(ComplexScenariosTests.ComplexScenariosTestData))>]
    member _.``Complex scenarios should parse correctly`` (testCase: TestCase) =
        // Arrange
        let input = testCase.Input
        let expectedJson = testCase.Expected

        // Act
        let result = parseDocument input
        let actualJson = toJson result
        let actualNode = JsonNode.Parse(actualJson)

        // Assert
        let expectedStr = expectedJson.ToJsonString()
        let actualStr = actualNode.ToJsonString()
        Assert.Equal(expectedStr, actualStr)

File: katas/kata-1/csharp/TimeToActDocumentAI.Tests/ComplexScenariosTests.cs
using System.Text.Json;
using System.Text.Json.Nodes;
using Xunit;

namespace TimeToActDocumentAI.Tests;

public class ComplexScenariosTests
{
    private static readonly JsonSerializerOptions JsonOptions = new()
    {
        PropertyNamingPolicy = JsonNamingPolicy.CamelCase,
        WriteIndented = true
    };

    [Theory]
    [MemberData(nameof(GetComplexScenarioTestCases))]
    public void ComplexScenarios_ShouldParseCorrectly(TestCase testCase)
    {
        // Arrange
        var input = testCase.Input;
        var expectedJson = testCase.Expected.ToJsonString();

        // Act
        var result = DocumentAI.ParseDocument(input);
        var actualJson = DocumentAI.ToJson(result);

        // Assert - Compare the JSON structures
        var expectedNode = JsonNode.Parse(expectedJson);
        var actualNode = JsonNode.Parse(actualJson);
        
        AssertJsonEquals(expectedNode, actualNode, testCase.Name);
    }

    public static IEnumerable<object[]> GetComplexScenarioTestCases()
    {
        return TestDataLoader.GetTestCases("complex-scenarios.json");
    }

    private static void AssertJsonEquals(JsonNode? expected, JsonNode? actual, string testName)
    {
        if (expected == null && actual == null) return;
        
        Assert.True(expected != null && actual != null, $"Null mismatch in test: {testName}");
        
        var expectedStr = expected.ToJsonString();
        var actualStr = actual.ToJsonString();
        
        Assert.Equal(expectedStr, actualStr);
    }
}